test/malloc_compliance_tests.cpp

```diff
@@ -0,0 +1,156 @@
+    }
+}
+
+//------------------------------------------------------------------------
+// Tests
+// -----------------------------------------------------------------------
+
+// ISO/IEC 9899:TC3 7.20.3.3
+void malloc_compliance_test(umf_memory_pool_handle_t hPool) {
+    void *alloc_ptr[ITERATIONS];
+    size_t alloc_ptr_size[ITERATIONS];
+
+    // Each allocation shall yield a pointer to an object disjoint from
+    // any other object.
+    for (int i = 0; i < ITERATIONS; i++) {
+        alloc_ptr_size[i] = rand_alloc_size(MAX_ALLOC_SIZE);
+        alloc_ptr[i] = umfPoolMalloc(hPool, alloc_ptr_size[i]);
+
+        // Supported under USE_WEAK_ALIGNMENT_RULES macro
+        // For compatibility, on 64-bit systems malloc should align to 16 bytes
+        size_t alignment =
+            (sizeof(intptr_t) > 4 && alloc_ptr_size[i] > 8) ? 16 : 8;
+        ASSERT_MSG(
+            is_aligned(alloc_ptr[i], alignment),
+            "Malloc should return pointer that is suitably aligned so that "
+            "it may be assigned to a pointer to any type of object ");
+
+        // Fill memory with zeros
+        ASSERT_MSG(alloc_ptr[i],
+                   "malloc returned NULL, couldn't allocate much memory");
+        memset(alloc_ptr[i], 0, alloc_ptr_size[i]);
+    }
+    for (int i = 0; i < ITERATIONS; i++) {
+        ASSERT_MSG(is_zero_mem(alloc_ptr[i], alloc_ptr_size[i]),
+                   "Object returned by malloc is not disjoined from others");
+        memset(alloc_ptr[i], 1, alloc_ptr_size[i]);
+    }
+    free_memory(hPool, alloc_ptr);
+}
+
+// ISO/IEC 9899:TC3 7.20.3.1
+void calloc_compliance_test(umf_memory_pool_handle_t hPool) {
+    void *alloc_ptr[ITERATIONS];
+    size_t alloc_size;
+
+    // Checking that the memory returned by calloc is zero filled
+    for (int i = 0; i < ITERATIONS; i++) {
+        alloc_size = rand_alloc_size(MAX_ALLOC_SIZE);
+        alloc_ptr[i] = umfPoolCalloc(hPool, i + 1, alloc_size);
+
+        ASSERT_MSG(alloc_ptr[i],
+                   "calloc returned NULL, couldn't allocate much memory");
+        ASSERT_MSG(is_zero_mem(alloc_ptr[i], alloc_size),
+                   "Memory returned by calloc was not zeroed");
+    }
+    free_memory(hPool, alloc_ptr);
+}
+
+// ISO/IEC 9899:TC3 7.20.3.4
+void realloc_compliance_test(umf_memory_pool_handle_t hPool) {
+    // If ptr is a null pointer, the realloc function behaves
+    // like the malloc function for the specified size
+    void *ret = umfPoolRealloc(hPool, NULL, 10);
+    ASSERT_MSG(ret, "If ptr is a NULL, realloc should behave like malloc");
+    // SIGSEGV if memory was allocated wrong
+    memset(ret, 1, 10);
+    umfPoolFree(hPool, ret);
+
+    // The contents of the new object shall be the same
+    // as that of the old object prior to deallocation
+    void *realloc_ptr[ITERATIONS];
+    size_t alloc_size;
+    for (int i = 0; i < ITERATIONS; i++) {
+        // Generate allocation size
+        alloc_size = rand_alloc_size(MAX_ALLOC_SIZE);
```

Comment:
I don't like adding randomness in tests. Can this be just a list of sizes?

Reply:
This list would have to have at least 100 (ITERATIONS) elements ... but ITERATIONS could be increased to 1000.
I have added `srand(SRAND_INIT_VALUE)` with constant `SRAND_INIT_VALUE` so it is almost a list of sizes now ;-)